pinot-common/src/main/java/org/apache/pinot/common/utils/SqlResultComparator.java

```diff
@@ -163,6 +167,97 @@ public static boolean areEqual(JsonNode actual, JsonNode expected, String query)
+
+    convertNumbersToString(expectedRows, columnDataTypes);
+    convertNumbersToString(actualRows, columnDataTypes);
+
+    List<String> actualElementsSerialized = new ArrayList<>();
+    List<String> expectedElementsSerialized = new ArrayList<>();
+    for (int i = 0; i < actualRows.size(); i++) {
+      actualElementsSerialized.add(actualRows.get(i).toString());
+    }
+    for (int i = 0; i < expectedRows.size(); i++) {
+      expectedElementsSerialized.add(expectedRows.get(i).toString());
+    }
+
+    if (!areLengthsEqual(actual, expected)) {
+      return false;
+    }
+
+    // For now, just directly compare elements in result set
+    if (!areNonOrderByQueryElementsEqual(actualElementsSerialized, expectedElementsSerialized)) {
+      return false;
+    }
+
+    // Compare stage stats
+    ObjectNode actualStageStats = (ObjectNode) actual.get(FIELD_MULTI_STAGE_STATS);
+    ObjectNode expectedStageStats = (ObjectNode) expected.get(FIELD_MULTI_STAGE_STATS);
+    return areMultiStageStatsEqual(actualStageStats, expectedStageStats);
+  }
+
+  private static boolean areMultiStageStatsEqual(ObjectNode actualStageStats, ObjectNode expectedStageStats) {
+    String actualType = actualStageStats.get(FIELD_MULTI_STAGE_STATS_TYPE) != null
+        ? actualStageStats.get(FIELD_MULTI_STAGE_STATS_TYPE).asText()
+        : null;
+    String expectedType = expectedStageStats.get(FIELD_MULTI_STAGE_STATS_TYPE) != null
+        ? expectedStageStats.get(FIELD_MULTI_STAGE_STATS_TYPE).asText()
+        : null;
+
+    if (actualType != null && !actualType.equals(expectedType)) {
+      LOGGER.error("Mismatch in stage stats type. Actual: {}, Expected: {}", actualType, expectedType);
+      return false;
+    }
+
+    ArrayNode actualChildren = (ArrayNode) actualStageStats.get(FIELD_MULTI_STAGE_STATS_CHILDREN);
+    ArrayNode expectedChildren = (ArrayNode) expectedStageStats.get(FIELD_MULTI_STAGE_STATS_CHILDREN);
+
+    if (actualChildren == null && expectedChildren == null) {
+      return true;
+    }
+    if (actualChildren == null) {
+      LOGGER.error("No children found in stage stats for type: {}. Expected {} children.",
+          actualType, expectedChildren.size());
+      return false;
+    }
+    if (expectedChildren == null) {
+      LOGGER.error("Found unexpected children in stage stats for type: {}. Expected no children.", actualType);
+      return false;
+    }
+    if (actualChildren.size() != expectedChildren.size()) {
+      LOGGER.error("Mismatch in number of children for stage stats for type: {}. Actual: {}, Expected: {}",
+          actualType, actualChildren.size(), expectedChildren.size());
+      return false;
+    }
+
+    for (int i = 0; i < actualChildren.size(); i++) {
+      if (!areMultiStageStatsEqual((ObjectNode) actualChildren.get(i), (ObjectNode) expectedChildren.get(i))) {
+        return false;
+      }
+    }
+
+    // TODO: Verify other stats like emittedRows, fanIn, fanOut, inMemoryMessages, rawMessages etc.?
```

Comment:
@gortiz what do you think? Which of these stats would be worthwhile to verify here?

Reply:
Right now our support for _backward compatibility_ means that the query should not fail. Therefore we may accept to lose stats. In future we may add support for stats as well, but let's not complicate the things too much in this first version

Reply:
Yeah, makes sense. I already ran into an issue when running this on a slightly older commit because of the change in the structure of multi-stage stats. Should I keep this as is or remove this stats check altogether?

Reply:
Discussed this offline, and we've decided to remove the multi-stage stats verification altogether since that is still subject to change across versions.